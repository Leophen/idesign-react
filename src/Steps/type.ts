export interface StepsItemProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 步骤标题
   * @default ''
   */
  title?: React.ReactNode;
  /**
   * 步骤描述
   * @default ''
   */
  description?: React.ReactNode;
  /**
   * index 值
   * @default 0
   */
  index?: number;
  /**
   * 进行到哪一步
   * @default 0
   */
  current?: number;
}

export interface StepsProps extends StepsItemProps {
  /**
   * 进行到哪一步
   * @default 0
   */
  current?: number;
  /**
   * 步骤条方向
   * @default horizontal
   */
  layout?: 'horizontal' | 'vertical';
  /**
   * 步骤条是否倒序
   * @default false
   */
  reverse?: boolean;
  /**
   * 是否为无序的步骤条
   * @default false
   */
  dot?: boolean;
}

export interface StepsContextValue {
  inject: (props: StepsProps) => StepsProps;
}
