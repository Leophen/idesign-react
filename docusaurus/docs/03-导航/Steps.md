---
title: Steps 步骤条
---

引导用户按照流程完成任务的步骤条。

## 引入

```js
import {Steps} from 'idesign';

const {Step} = Steps;
```

## 基本用法

当任务复杂或者存在先后关系时，将其分解成一系列步骤，从而简化任务。

```jsx live fffx
<Steps current={1}>
    <Step title="Finished" description="This is a description." />
    <Step title="In Progress" subTitle="Left 00:00:08" description="This is a description." />
    <Step title="Waiting" description="This is a description." />
</Steps>
```

## 迷你版

迷你版的步骤条，通过设置 `<Steps size="small">` 启用.

```jsx live fffx
<Steps size="small" current={1}>
    <Step title="Finished" />
    <Step title="In Progress" />
    <Step title="Waiting" />
</Steps>
```

## 横向步骤条上下布局方式

通过设置 `<Steps labelPlacement="vertical">` 启用

```jsx live fff
<Steps current={1} labelPlacement="vertical">
    <Step title="Finished" description="This is a description." />
    <Step title="In Progress" description="This is a description." />
    <Step title="Waiting" description="This is a description." />
</Steps>
```

## 步骤切换

通常配合内容及按钮使用，表示一个流程的处理进度。

```jsx live fff
function App() {
    const steps = [
        {
            title: 'First',
            content: 'First-content'
        },
        {
            title: 'Second',
            content: 'Second-content'
        },
        {
            title: 'Last',
            content: 'Last-content'
        }
    ];

    const [current, setCurrent] = React.useState(0);

    const next = () => {
        setCurrent(current + 1);
    };

    const prev = () => {
        setCurrent(current - 1);
    };

    return (
        <>
            <Steps current={current}>
                {steps.map(item => (
                    <Step key={item.title} title={item.title} />
                ))}
            </Steps>
            <div className="steps-content">{steps[current].content}</div>
            <div className="steps-action">
                {current < steps.length - 1 && (
                    <Button type="primary" onClick={() => next()}>
                      Next
                    </Button>
                )}
                {current === steps.length - 1 && (
                    <Button type="primary" onClick={() => alert('Processing complete!')}>
                      Done
                    </Button>
                )}
                {current > 0 && (
                    <Button style={{margin: '0 8px'}} onClick={() => prev()}>
                      Previous
                    </Button>
                )}
            </div>
        </>
    );
}
```

## 竖直方向的步骤条

简单的竖直方向的步骤条。

```jsx live fff
<Steps direction="vertical" current={1}>
    <Step title="Finished" description="This is a description." />
    <Step title="In Progress" description="This is a description." />
    <Step title="Waiting" description="This is a description." />
</Steps>
```

## 竖直方向的小型步骤条

简单的竖直方向的小型步骤条。

```jsx live fff
<Steps direction="vertical" size="small" current={1}>
    <Step title="Finished" description="This is a description." />
    <Step title="In Progress" description="This is a description." />
    <Step title="Waiting" description="This is a description." />
</Steps>
```

## 步骤运行错误

使用 Steps 的 `status` 属性来指定当前步骤的状态。

```jsx live fff
<Steps current={1} status="error">
    <Step title="Finished" description="This is a description" />
    <Step title="In Process" description="This is a description" />
    <Step title="Waiting" description="This is a description" />
</Steps>
```

## 点状步骤条

包含步骤点的进度条。

```jsx live fff
<div className="demo-column">
    <Steps progressDot current={1}>
        <Step title="Finished" description="This is a description." />
        <Step title="In Progress" description="This is a description." />
        <Step title="Waiting" description="This is a description." />
    </Steps>
    <Steps progressDot current={1} direction="vertical">
        <Step title="Finished" description="This is a description. This is a description." />
        <Step title="In Progress" description="error" status='error' />
        <Step title="Waiting" description="This is a description." />
    </Steps>
</div>
```

## 可点击的步骤条

设置 `onChange` 后，Steps 变为可点击状态。

```jsx live fff
function App() {
    const [current, setCurrent] = useState(0);

    const onChange = current => {
        console.log('onChange:', current);
        setCurrent(current);
    };

    return (
        <div className="demo-column">
            <Steps current={current} onChange={onChange}>
                <Step title="Step 1" description="This is a description." />
                <Step title="Step 2" description="This is a description." />
                <Step title="Step 3" description="This is a description." />
            </Steps>
            <Steps current={current} onChange={onChange} direction="vertical">
                <Step title="Step 1" description="This is a description." />
                <Step title="Step 2" description="This is a description." />
                <Step title="Step 3" description="This is a description - disabled" disabled />
            </Steps>
        </div>
    );
}
```

## API

### Steps

整体步骤条。

| 参数           | 说明                                                                          | 类型                                                                   | 默认值       |
| -------------- | ----------------------------------------------------------------------------- | ---------------------------------------------------------------------- | ------------ |
| className      | 步骤条类名                                                                    | `string`                                                               | `--`         |
| current        | 指定当前步骤，从 0 开始记数。在子 Step 元素中，可以通过 `status` 属性覆盖状态 | `number`                                                               | `0`          |
| direction      | 指定步骤条方向。目前支持水平（`horizontal`）和竖直（`vertical`）两种方向      | `string`                                                               | `horizontal` |
| initial        | 起始序号，从 0 开始记数                                                       | `number`                                                               | `0`          |
| labelPlacement | 指定标签放置位置，默认水平放图标右侧，可选 `vertical` 放图标下方              | `string`                                                               | `horizontal` |
| progressDot    | 点状步骤条，可以设置为一个 function，labelPlacement 将强制为 `vertical`       | `boolean〡(iconDot, {index, status, title, description}) => ReactNode` | `false`      |
| responsive     | 当屏幕宽度小于 532px 时自动变为垂直模式                                       | `boolean`                                                              | `--`         |
| size           | 指定大小，目前支持普通（`default`）和迷你（`small`）                          | `string`                                                               | `default`    |
| status         | 指定当前步骤的状态，可选 `wait` `process` `finish` `error`                    | `string`                                                               | `process`    |
| type           | 步骤条类型，有 `default` 和 `navigation` 两种                                 | `string`                                                               | `default`    |
| onChange       | 点击切换步骤时触发                                                            | `(current) => void`                                                    | `--`         |

### Steps.Step

步骤条内的每一个步骤。

| 参数        | 说明                                                                                                          | 类型        | 默认值  |
| ----------- | ------------------------------------------------------------------------------------------------------------- | ----------- | ------- |
| description | 步骤的详情描述，可选                                                                                          | `ReactNode` | `--`    |
| disabled    | 禁用点击                                                                                                      | `boolean`   | `false` |
| icon        | 步骤图标的类型，可选                                                                                          | `ReactNode` | `--`    |
| status      | 指定状态。当不配置该属性时，会使用 Steps 的 `current` 来自动指定状态。可选：`wait` `process` `finish` `error` | `string`    | `wait`  |
| subTitle    | 子标题                                                                                                        | `ReactNode` | `--`    |
| title       | 标题                                                                                                          | `ReactNode` | `--`    |
