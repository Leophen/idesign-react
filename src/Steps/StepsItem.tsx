import React, { useContext, useMemo } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import { StepsContext } from './index'
import { StepsItemProps } from './type';

const StepsItem: React.FC<StepsItemProps> = (props) => {
  const context = useContext(StepsContext);
  const newProps = context ? context.inject(props) : props;

  const {
    className,
    style,
    title = '',
    description = '',
    index = 0,
    current = 0,
    ...restProps
  } = newProps;

  return (
    <div
      className={classNames(
        'i-steps-item',
        current > index + 1 && 'i-steps-item__finish',
        current === index + 1 && 'i-steps-item__current',
        className
      )}
      style={{ ...style }}
      {...restProps}
    >
      <div className="i-steps-item__content">
        <header className="i-steps-item__title">
          {title}
          <div className="i-steps-item__icon">
            {current > index + 1 ? (
              <Icon name="Check" size={13} />
            ) : (
              <span>{index + 1}</span>
            )}
          </div>
        </header>
        <article className="i-steps-item__description">
          {description}
        </article>
      </div>
    </div>
  )
}

export default StepsItem;
