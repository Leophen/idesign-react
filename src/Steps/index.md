---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 16
---

# Steps 步骤条（开发中）

结合 element Timeline 时间线

## 基本用法

```tsx
import React from 'react';
import { Steps } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Steps>填充按钮</Steps>
      </div>
    </div>
  );
};

export default App;
```
