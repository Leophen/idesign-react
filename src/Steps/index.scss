
.i-steps {
  display: flex;
  &.i-steps__layout-vertical {
    flex-direction: column;
    .i-steps-item__title {
      &::after {
        display: none;
      }
    }
    .i-steps-item {
      padding-bottom: 20px;
      &:not(:last-of-type) {
        .i-steps-item__icon::after {
          content: '';
        }
      }
    }
    .i-steps-item__icon {
      &::after {
        position: absolute;
        top: calc(100% + 1px);
        left: 50%;
        display: block;
        width: 2px;
        height: 999px;
        background: var(--i-bg-back);
        transform: translateX(-50%);
      }
    }
  }
  &.i-steps__dot {
    .i-steps-item__content {
      margin-left: 20px;
    }
    .i-steps-item__icon {
      left: -20px;
      width: 12px;
      min-width: 12px;
      height: 12px;
      min-height: 12px;
      border-width: 2px;
      .i-icon,
      span {
        display: none;
      }
      &::after {
        top: calc(100% + 6px);
      }
    }
  }
}

.i-steps-item {
  display: flex;
  overflow: hidden;
  &:not(:last-of-type) {
    flex: 1;
    margin-right: 8px;
    .i-steps-item__title::after {
      content: '';
    }
  }
  &.i-steps-item__finish {
    &,
    .i-icon,
    .i-steps-item__icon {
      color: var(--i-primary-txt);
    }
    .i-steps-item__icon {
      border-color: var(--i-primary);
    }
    .i-steps-item__title {
      color: var(--i-font-2);
      &::after {
        background: var(--i-primary);
      }
    }
  }
  &.i-steps-item__current {
    &,
    .i-icon,
    .i-steps-item__icon {
      color: #fff;
    }
    .i-steps-item__icon {
      background: var(--i-primary);
      border-color: var(--i-primary);
    }
    .i-steps-item__title {
      color: var(--i-primary);
    }
  }
}

.i-steps-item__icon {
  position: absolute;
  top: 50%;
  left: -28px;
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: 20px;
  min-width: 20px;
  height: 20px;
  min-height: 20px;
  margin-right: 8px;
  color: var(--i-font-5);
  font-size: 12px;
  border: 1px solid var(--i-font-5);
  border-radius: 50%;
  transform: translateY(-50%);
  .i-icon {
    color: var(--i-font-5);
  }
}

.i-steps-item__content {
  margin-left: 28px;
}

.i-steps-item__title {
  position: relative;
  width: fit-content;
  color: var(--i-font-5);
  font-size: 14px;
  &::after {
    position: absolute;
    top: 50%;
    left: calc(100% + 6px);
    display: block;
    width: 999px;
    height: 2px;
    background: var(--i-bg-back);
    transform: translateY(-50%);
  }
}

.i-steps-item__description {
  margin-top: 8px;
  color: var(--i-font-5);
  font-size: 13px;
}
