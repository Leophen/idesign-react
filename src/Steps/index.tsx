import React, { useMemo } from 'react';
import classNames from 'classnames';
import './index.scss';
import StepsItem from './StepsItem';
import { StepsContextValue, StepsProps } from './type';

export const StepsContext = React.createContext<StepsContextValue>(null as any);

const Steps: React.FC<StepsProps> & { Item: React.ElementType } = (props) => {
  const {
    children = '',
    className,
    style,
    current = 0,
    layout = 'horizontal',
    reverse,
    dot = false
  } = props;

  // 注入每一项的 context
  const context: StepsContextValue = {
    // 将 props 注入每一项子节点的方法
    inject: (stepsItemProps: StepsProps) => {
      return {
        current,
        ...stepsItemProps,
      };
    },
  };

  const stepItemList = useMemo(() => {
    const childrenList = React.Children.toArray(children);
    const childrenDisplayList = reverse ? childrenList.reverse() : childrenList;

    return childrenList.map((child: any, index: number) => {
      const stepIndex = reverse ? childrenDisplayList.length - index - 1 : index;
      return React.cloneElement(child, {
        ...child.props,
        index: stepIndex,
      });
    });
  }, [children, reverse]);

  return (
    <StepsContext.Provider value={context}>
      <div
        className={classNames(
          'i-steps',
          layout && `i-steps__layout-${layout}`,
          dot && 'i-steps__dot',
          className
        )}
        style={{ ...style }}
      >
        {stepItemList}
      </div>
    </StepsContext.Provider>
  );
};

Steps.Item = StepsItem

StepsItem.displayName = 'StepsItem';
Steps.displayName = 'Steps';

export default Steps;
