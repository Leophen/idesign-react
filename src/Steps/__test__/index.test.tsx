import { render } from '@testing-library/react';
import Steps from '../index';

describe('Steps 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Steps>
        <Steps.Item
          title="步骤1"
          description="提示文字"
        />
        <Steps.Item
          title="步骤2"
          description="提示文字"
        />
        <Steps.Item
          title="步骤3"
          description="提示文字"
        />
      </Steps>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('current', () => {
    const { getByTestId } = render(
      <Steps current={2}>
        <Steps.Item
          title="步骤1"
          description="提示文字"
          data-testid='first-item'
        />
        <Steps.Item
          title="步骤2"
          description="提示文字"
          data-testid='second-item'
        />
        <Steps.Item
          title="步骤3"
          description="提示文字"
        />
      </Steps>
    );
    expect(getByTestId('first-item')).toHaveClass('i-steps-item__finish');
    expect(getByTestId('second-item')).toHaveClass('i-steps-item__current');
  });

  it('dot', () => {
    const { container } = render(
      <Steps dot>
        <Steps.Item
          title="步骤1"
          description="提示文字"
        />
        <Steps.Item
          title="步骤2"
          description="提示文字"
        />
        <Steps.Item
          title="步骤3"
          description="提示文字"
        />
      </Steps>
    );
    expect(container.firstChild).toHaveClass('i-steps__dot');
  });

  it('layout', () => {
    const { container } = render(
      <Steps layout="vertical">
        <Steps.Item
          title="步骤1"
          description="提示文字"
        />
        <Steps.Item
          title="步骤2"
          description="提示文字"
        />
        <Steps.Item
          title="步骤3"
          description="提示文字"
        />
      </Steps>
    );
    expect(container.firstChild).toHaveClass('i-steps__layout-vertical');
  });

  it('reverse', () => {
    const { container, getByTestId } = render(
      <Steps reverse>
        <Steps.Item
          title="步骤1"
          description="提示文字"
        />
        <Steps.Item
          title="步骤2"
          description="提示文字"
        />
        <Steps.Item
          title="步骤3"
          description="提示文字"
          data-testid='test-item'
        />
      </Steps>
    );
    expect(container.firstChild.firstChild).toEqual(getByTestId('test-item'));
  });
});

