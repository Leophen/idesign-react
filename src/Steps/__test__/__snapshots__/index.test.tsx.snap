// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-steps i-steps__layout-horizontal"
  >
    <div
      class="i-steps-item"
    >
      <div
        class="i-steps-item__content"
      >
        <header
          class="i-steps-item__title"
        >
          步骤1
          <div
            class="i-steps-item__icon"
          >
            <span>
              1
            </span>
          </div>
        </header>
        <article
          class="i-steps-item__description"
        >
          提示文字
        </article>
      </div>
    </div>
    <div
      class="i-steps-item"
    >
      <div
        class="i-steps-item__content"
      >
        <header
          class="i-steps-item__title"
        >
          步骤2
          <div
            class="i-steps-item__icon"
          >
            <span>
              2
            </span>
          </div>
        </header>
        <article
          class="i-steps-item__description"
        >
          提示文字
        </article>
      </div>
    </div>
    <div
      class="i-steps-item"
    >
      <div
        class="i-steps-item__content"
      >
        <header
          class="i-steps-item__title"
        >
          步骤3
          <div
            class="i-steps-item__icon"
          >
            <span>
              3
            </span>
          </div>
        </header>
        <article
          class="i-steps-item__description"
        >
          提示文字
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
